pkg/api/apierror/error.go

```diff
@@ -0,0 +1,209 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package apierror
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"reflect"
+	"unsafe"
+
+	"github.com/go-openapi/runtime"
+
+	"github.com/elastic/cloud-sdk-go/pkg/models"
+	"github.com/elastic/cloud-sdk-go/pkg/multierror"
+)
+
+const (
+	// ErrTimedOutMsg is returned when the error is context.DeadlineExceeded.
+	ErrTimedOutMsg = "operation timed out"
+
+	errPrefix = "api error"
+)
+
+type knownError interface {
+	GetPayload() *models.BasicFailedReply
+}
+
+// Error wraps a API error and implements both the Error and Unwrap interfaces.
+type Error struct {
+	// Err is the unmodified wrapped error.
+	Err error
+
+	// merr is returned in the case that multiple errors are found as a
+	// *models.BasicFailedReply.
+	merr *multierror.Prefixed
+
+	// err will only be populated when merr is not.
+	err error
+}
+
+// Wrap creates a new Error from the passed error, it doesn't modify the
+// original error, but instead, tries to unwrap and extract the wrapped
+// errors form the go-openapi/runtime operations. The original error type
+// can be accessed by calling Unwrap() or errors.Unwrap/Is/As.
+//
+// If the error is of *models.BasicFailedReply type, then the errors are
+// unpacked into a *multierror.Prefixed which can be obtained by calling
+// Multierror().
+// The other possible case is when the API returns an error with an unexpected
+// status code which is not present in the swagger spec, in which case,
+// the same operation will be attempted (unwrapping a BasicFailedReply)
+// with a fallback to unmarshal any JSON which can be read:
+// * error is nil, in which case nil is returned.
+// * error is a context.DeadlineExceeded error, which equals a timeout.
+// * error is of type *runtime.APIError, meaning the returned API error wasn't
+//   defined in the Swagger spec from which the source code has been generated
+//   * HTTP code is 449, the authenticated user needs to elevate-permissions.
+//   * The type wraps *http.Response, the body is read and tries json.Unmarshal
+//     to *models.BasicFailedResponse and each of the BasicFailedReplyElement
+//     is then added to an instance of multierror.Prefixed and returned.
+//   * The error is unknown, returns "<OperationName> (status <StatusCode)".
+func Wrap(err error) error {
+	if err == nil {
+		return nil
+	}
+
+	result := Error{Err: err}
+
+	if bfr, ok := err.(knownError); ok {
+		m := newBasicFailedReplyMultierror(errPrefix, bfr.GetPayload())
+		if m.ErrorOrNil() != nil {
+			result.merr = m
+			return &result
```

Comment:
We are returning a pointer, I was using the pointer semantics that make it easy to determine the return type of the variable just by reading the code, in each of the returns you can always see that because `return &result` is used, that it is a pointer of the type `Error` that is returned.

If I initialize the `result` variable and set it to a pointer which is not necessary in the boundary of this function, then all returns do not tell you if we're returning a pointer or we're returning a normal value.